### ISSUE TITLE
Including trailing slashes in 

### ISSUE THREAD

**Issue by cbp123** [git-for-windows/git#1444]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.16.0.windows.2
cpu: x86_64
built from commit: a56c4f9e2a9513d2be57dc4346e0e84887399cd9
sizeof-long: 4

```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
Microsoft Windows [Version 10.0.16299.192]

** insert your machine's response here **
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VIM
Path Option: Cmd
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Installed via chocolatey.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Powershell

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```

1 d:
2 mkdir Projects
3 cd Projects
4 mkdir "My Project A"
5 mkdir "My Project B"
6 cd '.\My Project A\'
7 git init
8 new-item test.txt
9 git add .
10 git commit -m "Test"
11 cd ..
12 cd '.\My Project B\'
13 git clone --bare "D:\Projects\My Project A\" .\
14 git clone --bare "D:\Projects\My Project A" .\
15 cd ..
16 cd '.\My Project A\'
17 git remote add test 'D:\Projects\My Project B\'
18 git remote -v
19 git push test master

```
 - What did you expect to occur after running these commands?

 - What actually happened instead?

At step 13 we see the first incident of confusing behaviour, with `git clone` return an `invalid argument` message. For a user, it is quite challenging to figure out that the trailing slash should be removed.

A bigger problem begins at step 17. `git remote add` inexplicably adds a trailing double quote to the remote repository's URL. Again, the solution of removing the trailing slash is obscure.

This behaviour is annoying, because Powershell inserts the trailing slash by default when 'tabbing' a directory.

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

N/A

**Comment by shiftkey** [comment]
@cbp123 is this something that has worked for you before when using Git on Windows?

**Comment by cbp123** [comment]
Hi @shiftkey, I haven't tried this particular scenario before (remoting to repositories on the same local computer).

**Comment by dscho** [comment]
> This behaviour is annoying, because Powershell inserts the trailing slash by default when 'tabbing' a directory.

I totally agree, yet I do not really know what we can do about it. You see, when you call your instruction 17 (_very_ good MCVE, BTW, thank you so much, it is refreshing!), the command line sent to `git.exe` is this:

```
cmdline: '"C:\Program Files\Git\mingw64\bin\git.exe" remote add test2 "D:\Projects\My Project B\"'
```

(Note: there is a subtlety here that will be important later: that `git.exe` is not called directly, but via `C:\Program Files\Git\cmd\git.exe`: `C:\Program Files\Git\cmd` is in the `PATH`, to keep the search path clean from all those Unix-y tools that Git requires in its own `PATH`.)

Now, `git.exe` is written in portable C89/C90, meaning that it does not parse this command line, but expects it as a counted string array (`int main(int argc, const char **argv`). And once that parsing was done, we see this:

```
arg #0: 'C:\Program Files\Git\mingw64\bin\git.exe'
arg #1: 'remote'
arg #2: 'add'
arg #3: 'test2'
arg #4: 'D:\Projects\My Project B"'
```

(The single quotes are an artifact of my debug print statements, and not part of the actual elements of the `argv` array.)

Clearly, the problem is that the `\"` was shortened to `"`.

However, this is *actually* as intended. Because, you see, `git.exe` is not only written as a portable C89/C90 program, but it is meant to be executed from a Unix shell. (Yes, yes, we do not have a native Unix shell on Windows, but Git was not invented to run on Windows, Git was invented to supersede BitKeeper as the version control system of the Linux kernel project, and while Git has undoubtedly outgrown its humble beginnings, there are still remainders that show their heads every once in a while).

And in a Unix shell, the backslash is the escape character. That is, if you want to pass a literal double quote to a program in Bash (as opposed to using a double quote to pass a command-line argument containing e.g. spaces), you prefix it with a backslash.

Example: `echo "git"` will output `git`. `echo \"git` will output `"git`.

Typically, the Unix shell already performs this kind of command-line parsing itself. But on Windows, you do not call `execv(path, argv)`, not really, you call `CreateProcessW()`, which passes the *entire* command-line to the spawned process. Therefore, it cannot be the calling process that parses the command-line, it is always the called process. That's just how things are on Windows, and the Git project was started on a Unix-y system where command-line parsing is done by the caller, and that's that. On Windows, therefore, `git.exe` *has* to parse the command-line. And since it is most typically called from a Bash (it even calls itself via Bash, so there's that), it uses the Unix shell parsing rules for the command line.

That means `git.exe` does the exact right thing, but PowerShell does not know that this is a Unix shell program that tries to do its best on Windows. And therefore it happily passes a command-line that it thinks will be parsed according to Windows rules where, let's admit it, quoting rules are kind of a mess.

So where does this leave us? I don't quite know...

I guess what we *could* do is to install a `git.ps1` in `C:\Program Files\Git\cmd` that intercepts `git` calls, parses the command-line in the intended way, then re-quotes it in a way expected by `git.exe` and hands off to the real `git.exe`.

If you would be happy with that solution, I would ask you to try your hand at it because your PowerShell fu seems to be about 1e32 times better than mine 😄

**Comment by cbp123** [comment]
Thanks for the detailed **dscho**.
Assuming the general architecture stays the same, it sounds like some sort of adapter would be the logical way to go. Happy to get the ball rolling. Let me know if you think of any other alternatives, but in the mean time will see how I go with a Powershell adapter...

**Comment by dscho** [comment]
> will see how I go with a Powershell adapter as you suggest...

Please keep me posted. I will be very curious to see this happening.

Alternatively, if you have an idea how to teach the Git wrapper in `C:\Program Files\Git\cmd` (which is written in C) to detect that it was called from PowerShell and/or CMD, and to re-quote the command-line according to the MINGW rules that `C:\Program Files\Git\mingw64\bin\git.exe` expects, I would be *very* interested, too.

**Comment by cbp123** [comment]
I tried intercepting the call to `C:\Programs Files\Git\cmd\git.exe` with a .ps1 file, as you suggested (quite clever), and it does work... kind of. On the one hand, this solution feels nice because we can add a decoupled "Powershell adapter", rather than trying to detect the caller from inside git.exe. On the other hand, the solution feels troublesome and hacky. We are relying on a quirk of Powershell prioritising the `.ps1` over `.exe`. Also, the solution doesn't work for CMD prompt. And unfortunately when the extension is not specified, by default `.exe` files take priority over `.bat` or `.cmd` file, so we can't intercept using a `.bat` or `.cmd`. Possibly something could be done with a `.com` file, but `.com` won't fire by default on a 64-bit system.

There are, of course, more radical approaches, such as renaming git.exe, but right now I'm thinking your second suggestion is the better way - to tackle the C code in `C:\Programs Files\Git\cmd\git.exe`. That's a bit of a bigger job for me, but I'll give it a shot and let you know if I can take ownership. In the meantime, we could perhaps use the Powershell .ps1 trick, because at least that alleviates the issue in Powershell.

**Comment by rimrul** [comment]
>Possibly something could be done with a .com file, which takes priority over .exe, but .com won't fire by default on a 64-bit system.

That isn't quite true. While `.com` and `.exe` originally refered to different styles of executables, semi-modern Windows versions regard those extensions as interchangable. By default `CMD` still prefers `.com` files over `.exe`, but a `.com` file can still be in `PE` format and run on 64-bit systems.

See also:
https://devblogs.microsoft.com/oldnewthing/20080324-00/?p=23033
https://en.wikipedia.org/wiki/COM_file#Use_for_compatibility_reasons

**Comment by dscho** [comment]
I figure that a hacky way would be to use [the somewhat hacky `GetParentProcessId()` function](https://stackoverflow.com/a/3137081) and then retrieve the command line of said process to determine whether it is PowerShell or CMD...

**Comment by dscho** [comment]
(Or maybe it should actually test for `sh.exe` or `bash.exe`?)

**Comment by cbp123** [comment]
I've put some further research into this.

Firstly, here is some information about the weirdness of backslashes in Windows: https://weblogs.asp.net/jongalloway/_5B002E00_NET-Gotcha_5D00_-Commandline-args-ending-in-_5C002200_-are-subject-to-CommandLineToArgvW-whackiness

But it gets even weirder since that article, because nowadays the behaviour is actually different between CMD and Powershell. For example, if we pass arguments to a simple C# console app like the one below, we get different behaviour in CMD and Powershell:

```
using System;
namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            foreach (var arg in args)
            {
                Console.WriteLine(arg);
            }
        }
    }
}
```

In CMD:

```
D:\MyApp\MyApp.exe "C:\Projects\"
Hello World!
C:\Projects"
```

In Powershell:
```
D:\MyApp\MyApp.exe "C:\Projects\"
Hello World!
C:\Projects\
```

Annoyingly, the equivalent program written in C exhibits _the same_ behaviour in both CMD and Powershell! So clearly .NET and Powershell have been re-designed to play well together, but of course, now life is even more confusing...

In conclusion, it's still not clear to me if there is a simple solution to the problem, but I'll keep thinking. I do know that we can intercept calls to git.exe by creating a git.com file. This gives a number of options for a wrapper (that could potentially be written in .NET, if there is some way of cleaning up the arguments).

**Comment by dscho** [comment]
Thank you for continuing this investigation!

I still think that going for a `git.ps1` wouldn't be _too_ bad. In particular because it could imitate our old `git.cmd` which was the precursor to the Git wrapper (`/cmd/git.exe`): it simply sets a couple of environment variables (`MSYSTEM=MINGW64`, `PATH=..\mingw64\bin;..\usr\bin;%PATH%` are the ones that come to mind) and then hands off to `mingw64\bin\git.exe`.

This should be pretty easy to replicate in PowerShell, and also to add a simple adapter that figures out whether arguments refer to absolute paths and converts them before handing them off to `git.exe`...

**Comment by cbp123** [comment]
The following Powershell script (cmd\git.ps1) should do the job, above caveats aside.

`
$argList = ""

foreach($arg in $args)
{	
	# Skip empty arguments
	if ($arg -eq "") {
		continue
	}
	
	# Escape backslashes and double quotes (using a syntax that will work in bash)
	$parsedArg = $arg.Replace("\", "\\").Replace("`"", "\```"")
	
	# Put double quotes around arguments that contain spaces
	if ($parsedArg.Contains(" "))
	{
		$parsedArg = "`"" + $parsedArg + "`""
	}
	
	# Concatenate parsed argument
	if ($argList -ne "") {
		$argList += " "
	}
		
	$argList += $parsedArg
}

Invoke-Expression "git.exe $argList"
`

Happy for you to take this from here, or are you able to point me in the right direction for integrating into the code-base at the right spot (I've pulled down git-sdk-64 but still a bit overwhelmed)? And if necessary on submitting the appropriate pull request?

**Comment by cbp123** [comment]
The following Powershell script (cmd\git.ps1) should do the job, above caveats aside.

~~~~
$argList = ""

foreach($arg in $args)
{	
	# Skip empty arguments
	if ($arg -eq "") {
		continue
	}
	
	# Escape backslashes and double quotes (using a syntax that will work in bash)
	$parsedArg = $arg.Replace("\", "\\").Replace("`"", "\```"")
	
	# Put double quotes around arguments that contain spaces
	if ($parsedArg.Contains(" "))
	{
		$parsedArg = "`"" + $parsedArg + "`""
	}
	
	# Concatenate parsed argument
	if ($argList -ne "") {
		$argList += " "
	}
		
	$argList += $parsedArg
}

Invoke-Expression "git.exe $argList"
~~~~

Happy for you to take this from here, or are you able to point me in the right direction for integrating into the code-base at the right spot (I've pulled down git-sdk-64 but still a bit overwhelmed)? And if necessary on submitting the appropriate pull request?

**Comment by dscho** [comment]
> are you able to point me in the right direction for integrating into the code-base at the right spot

Absolutely. This should live next to `cmd\start-ssh-agent.cmd`, which is tracked in https://github.com/git-for-windows/MINGW-packages/tree/master/mingw-w64-git. This `.ps1` should probably be handled the same way.

Specifically, please
- add it to the `mingw-w64-git` subdirectory,
- add it to [the `source` section in the `PKGBUILD` file](https://github.com/git-for-windows/MINGW-packages/blob/ca39112339c8731315514fd1cf05c6840c5d4e5d/mingw-w64-git/PKGBUILD#L47) in the same directory,
- run `updpkgsums` to update [the `sha256sums`](https://github.com/git-for-windows/MINGW-packages/blob/ca39112339c8731315514fd1cf05c6840c5d4e5d/mingw-w64-git/PKGBUILD#L60)
- add it to the [`install` command in the `package_git()` function](https://github.com/git-for-windows/MINGW-packages/blob/ca39112339c8731315514fd1cf05c6840c5d4e5d/mingw-w64-git/PKGBUILD#L184) so that it gets copied into the correct location.

To test this, build the package with `MAKEFLAGS=-j8 makepkg-mingw` and verify that it is part of the generated package.

Then, open a PR at https://github.com/git-for-windows/MINGW-packages.

Thank you!

**Comment by cbp123** [comment]
Hm, I've discovered an issue with the above solution which I don't understand as yet.

It seems that the commit message argument (i.e. git commit -m) is processed without any escaping of backslashes, so it behaves differently to other arguments.

For example:
`
git commit -m "Test\"
git log`

The backslash is not treated as the escape character for the double quote.
Unlike in the following case:
`
git clone "D:\Project\My Project\"
`

I'm not sure what causes this differences yet, whether there is a specific handling of the commit message argument in the code base, or whether something else is at play.

**Comment by dscho** [comment]
IIRC the unquoting rules in CMD's command-line "REPL" are a bit funny. Unless my memory fails me, if there is more than one double quote, the outer-most two double quotes are simply stripped before passing on the command-line to the executable. Sounds like you were as surprised about this behavior as I was.

**Comment by dscho** [comment]
Closing this as stale.
